src/features/weak-references.md

```diff
@@ -54,104 +54,133 @@ A JavaScript `WeakMap` is not really _weak_: it actually refers _strongly_ to it
 `WeakRef` is a more advanced API that provides _actual_ weak references, enabling a window into the lifetime of an object. Let’s walk through an example together.
 
-Imagine a `getImage` function that takes a `name` and performs some expensive operation to generate another object, like a binary blob of image data:
+For the example, suppose we're working on a chat web application that uses web sockets to communicate with a server. Imagine a `MovingAvg` class that, for performance diagnostic purposes, keeps a set of events from a web socket in order to compute a simple moving average of the latency.
 
 ```js
-function getImage(name) {
-  const image = performExpensiveOperation(name);
-  return image;
+class MovingAvg {
+  constructor(socket) {
+    this.events = [];
+    this.socket = socket;
+    this.listener = (ev) => { this.events.push(ev); };
+    socket.addEventListener('message', this.listener);
+  }
+
+  compute(n) {
+    // Compute the simple moving average for the last n events.
+    // ...
+  }
 }
 ```
 
-To improve performance, we store the image in a cache. Now, we don’t have to perform the expensive operation again for the same name!
+It is used by a `MovingAvgComponent` class that lets you control when to start and stop monitoring the simple moving average of the latency.
 
 ```js
-const cache = new Map();
+class MovingAvgComponent {
+  constructor(socket) {
+    this.socket = socket;
+  }
+
+  start() {
+    this.movingAvg = new MovingAvg(this.socket);
+  }
 
-function getImageCached(name) {
-  if (cache.has(name)) return cache.get(name);
-  const image = performExpensiveOperation(name);
-  cache.set(name, image);
-  return image;
+  stop() {
+    // Allow the garbage collector to reclaim memory.
+    this.movingAvg = null;
+  }
+
+  render() {
+    // Do rendering.
+    // ...
+  }
 }
 ```
 
-But, there’s a problem here. `Map`s hold on to their keys and values strongly, and so the image names and data can never be garbage-collected. This steadily increases memory and causes a **memory leak**!
+We know that keeping all the server messages inside an instance `MovingAvg` uses a lot of memory, so we take care to null out `this.movingAvg` when monitoring is stopped to let the garbage collector reclaim memory.
 
-`WeakRef`s make it possible to solve the memory leak by creating a _weak reference_ to the image and storing _that_ in the cache (instead of the image itself). This way, the garbage collector can clean up the images that do not have a strong reference.
+However, after checking in the memory panel in DevTools, we found out that memory was not being reclaimed at all! The seasoned web developer may have already spotted the bug: event listeners are strong references and must be explicitly removed. Thus, the listener in `MovingAvg` instances, by referencing `this`, keeps the whole instance alive as long as the event listener isn't removed.
+
+Until now, the solution is to manually unregister the event listener via `dispose` method.
```

Comment:
```suggestion
Until now, the solution is to manually unregister the event listener via the `dispose` method.
```

Reply:
I would say "via **a** dispose method"

Reply:
Right, that's better indeed! Edited the suggestion

Reply:
Done.